STIPS/STIP-004.md

```diff
@@ -0,0 +1,730 @@
+    -  Current approach is leaning towards not restricting module usage but providing it as a potential different offering for manager's to choose when selecting the issuance module they want to use
+    - They can choose the default issuance modules with stricter checks if they are sure of not using any aToken or tokens which take fees on transfer as components in their SetToken
+    - But if they plan to use aTokens or tokens which take fees on transfer as SetToken components, then they can use the new issuance modules
+
+
+## Checkpoint 1
+
+**Reviewer**:
+
+## Proposed Architecture Changes
+### IssuanceUtils 
+A library contract containing a collection of utility functions to help during issuance/redemption of SetToken. This contract would contain the proposed collateralization checks and would be called from issuance modules.
+
+#### Methods
+
+- The methods are categorized into validate collateralization after components are transferred
+   - IN to the SetToken during issuance/redemption
+   - OUT of the SetToken during issuance/redemption
+- On transfer IN, we ensure collateralization by making sure the new component balance is greater than the sum of the `required` existing balance and component quantity transferred in.
+- On transfer OUT, we ensure collateralization by making sure the new component balance is greater than the `required` balance considering the new SetToken supply.
+
+
+> **validateCollateralizationPostTransferInPreHook**
+* Checks for *newBalance >= s * defaultPositionUnit + _componentQuantity*, where *s* is token supply before issuance/redemption. 
+```solidity
+/**
+ * Validates component transfer IN to SetToken during issuance/redemption. Reverts if Set is undercollateralized post transfer.
+ * NOTE: Call this function immediately after transfer IN but before calling external hooks (if any).
+ *
+ * @param _setToken             Instance of the SetToken being issued/redeemed
+ * @param _component            Address of component being transferred in/out
+ * @param _componentQuantity    Amount of component transferred into SetToken
+ * @param _isIssue              True if issuing SetToken, false if redeeming
+ * @param _setQuantity          Amount of SetToken burnt in this transaction. If being issued, pass 0.
+ *                              This value is used to calculate the current SetToken supply.
+ */
+function validateCollateralizationPostTransferInPreHook(
+   ISetToken _setToken, 
+   address _component, 
+   uint256 _componentQuantity, 
+   bool _isIssue, 
+   uint256 _setQuantity
+) 
+   internal view
+```
+
+* Called in *DebtIssuanceModuleV2#_resolveEquityPositions* during issuance after each equity is transferred in but before external component hooks are called.
+   * Check we devised above: _require(newBalance >= s * defaultPositionUnit + i * cumulativeEquity)_
```

Comment:
For max clarity can we state how we're calculating `s` for each of these scenarios. I'd think here we'd want to check that `totalSupply + issueAmount` is held by the Set. But seems we're passing in `0` to `_setQuantity`